CallerEnvironment type

stepancheg opened the pull request:
Caller environment can only be used to revolve "type values" and
nothing more.

It might be convenient to access true caller environment, but it leads
to certain problems. Consider this situation. Module `a.bzl`
defines a function `foo` which calls function `bar`.

```
def bar(): ...

def foo(): bar()
```

and `BUILD` file imports `a.bzl` and calls `foo`:

```
foo()
```

`foo` function captures the context of `a.bzl` and inherits it in
locals (since #113).

However, when `foo` resolves type values, it must use type values
defined in the root caller (`BUILD` context, not captured).

Now function `foo` calls `bar`. Function `bar` also needs to resolve
type values.  These type values must again come from `BUILD` (because
captured context might not contain proper type values).

This environment passed to `bar` must be `BUILD` environment, but
not the local environment of `foo`.

Why the environment passed to `bar` must be `BUILD` environment?
Consider this scenario, `baz` is defined in `baz.bzl`:

```
def baz(unknown): unknown.append(10)
```

When `baz` is called from `BUILD` with object `foo` defined elsewhere,
`baz` should use the caller context, not local or captured context
to resolve `unknown.bar`. Because local (and captured) context might
not contain type values for `unknown`. But the function definition
is perfectly valid.

In theory in `foo` when calling `bar` we could merge the local
environment with the environment from `BUILD`, but:

* it is not practically needed for anything
* it complicates everything beyond the limit of understanding by human

`CallerEnvironment` type is introduced which wraps `Environment`
but only provides access to type values. And `Environment` no longer
provides readable access to type values.

The good side is that change forced me to fix a couple of issues
with using an incorrect environment to find type values.

The bad side is that `dbg` function is no longer able to print
caller locals.

But the good side is that `dbg` was broken anyway after fix #113.

damienmg commented:
/cc @laurentlb @adonovan: is this correct that a function from a different module see a global values from the caller environment, e.g.: 

def.bzl:
```python
def fn(): return unknown + [10]
```

BUILD:
```python
load(":def.bzl", "fn")

unknown = [1,2,3]

known = fn()
````

alandonovan commented:
> is it correct that a function from a different module sees global values from the caller environment?

No, Starlark is statically scoped, like all sane languages in the last 50 years.

Look at the Go code if you want to see how to implement it. When we create the implicit "toplevel" function for all the toplevel statements of module, we create the set of globals (and constants and predeclared values) for the module:
https://github.com/google/starlark-go/blob/master/starlark/eval.go#L387
and we create any by a def statement or lambda expression, we share the set of globals/constants/predeclared values with the enclosing function:
https://github.com/google/starlark-go/blob/master/starlark/interp.go#L479
Thus all functions (and toplevel statements) in the same module see the same set of globals.

An upcoming refactoring in https://github.com/google/starlark-go/pull/203 introduces an (internal) "module" data type that makes the structure even more explicit.

damienmg commented:
We are in violent agreement, I think I misread the explanation of this PR. @stepancheg: can you add test that this is doing the right thing? (e.g. defining unknown at various place and failing or suceeding)

stepancheg commented:
* renamed `CallerEnvironment` to `TypeValues`
* added a test which verifies where "type values" come from
* existing `test_context_captured` verifies that globals (which are not type values) are captured, not obtained from caller context

Note semantic of captures was changed and fixed in #113, this PR just makes it more explicit and type safe.
